DESIGN.md

```diff
@@ -0,0 +1,773 @@
+feature is adding new types to represent the registered state. For example,
+`File` could have a `RegisteredFile` analog. Making this change would be
+forwards compatible and not impact the current design.
+
+## Alternatives
+
+### Use a work-stealing scheduler
+
+An alternative approach could use a work-stealing scheduler, allowing
+underutilized worker threads to steal work from overloaded workers. While
+work-stealing is an efficient strategy for balancing load across worker threads,
+required synchronization adds overhead. [An earlier article] on the Tokio blog
+includes an overview of various scheduling strategies.
+
+The tokio-uring crate targets use-cases that can benefit from taking advantage
+of io-uring at the expense of discarding Tokio's portable API. These use cases
+will also benefit from reduced synchronization overhead and fine-grained control
+over thread load balancing strategies.
+
+[An earlier article]: https://tokio.rs/blog/2019-10-scheduler
+
+### Read and write methods on TcpStream
+
+Read and write operations on byte stream types, such as `TcpStream`, are
+stateful. Each operation operates on the next chunk of the stream, advancing a
+logical cursor. The combination of submit-based operations with drop-to-cancel
+semantics presents a challenge. Consider the following.
+
+```rust
+let buf = IoBuf::with_capacity(4096);
+
+// This read will be canceled.
+select! {
+    _ = my_tcp_stream.read(buf.slice(..)) => unreachable!(),
+    _ = future::ready(()) => {}
+}
+
+let buf = IoBuf::with_capacity(4096);
+let (res, buf) = my_tcp_stream.read(buf.slice(..)).await;
+```
+
+Dropping the first read operation cancels it, and because it never completes,
+the second read operation should read the first packet of data from the TCP
+stream. However, the kernel may have already completed the first read operation
+before seeing the cancellation request. A naive runtime implementation would
+drop the completion result of any canceled operation, which would result in
+losing data. Instead, the runtime could preserve the data from the first read
+operation and return it as part of the second read. The process would not lose
+data, but the runtime would need to perform an extra copy or return the caller a
+different buffer than the one it submitted.
+
+
+The proposed API is not vulnerable to this issue as resources track their
+operations, preventing them from being dropped as long as the resource is open.
+
+### Expose a raw io-uring operation submission API
+
+The proposed tokio-uring API does not include a strategy for the user to submit
+custom io-uring operations. Any raw API would be unsafe or would require the
+runtime to support taking opaque data as a trait object. Given that io-uring has
+a well-defined set of supported options, tokio-uring opts to support each
+operation explicitly. The application also can create its own set of io-uring
+queues using the io-uring crate directly.
+
+### Load-balancing spawn function
```

Comment:
About load balance, If we do clientside load balancing, the server tokio-uring should support which feature?   We should mention is .